src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -231,101 +216,164 @@ private static bool CanOptimizeSingleSpreadAsCollectionBuilderArgument(BoundColl
+                    => createImmutableArray(),
+                _ => throw ExceptionUtilities.UnexpectedValue(collectionType),
+            };
+
+            ArrayTypeSymbol getBackingArrayType()
+            {
+                Debug.Assert(collectionType.OriginalDefinition == (object)_compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T) ||
+                    collectionType.OriginalDefinition == (object)_compilation.GetWellKnownType(WellKnownType.System_Span_T) ||
+                    collectionType.OriginalDefinition == (object)_compilation.GetWellKnownType(WellKnownType.System_Collections_Immutable_ImmutableArray_T));
+
+                var elementType = ((NamedTypeSymbol)collectionType).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0];
+                return ArrayTypeSymbol.CreateSZArray(_compilation.Assembly, elementType);
+            }
 
-            var arrayType = collectionType as ArrayTypeSymbol;
-            if (arrayType is null)
+            BoundExpression createImmutableArray()
             {
-                // We're constructing a Span<T> or ReadOnlySpan<T> rather than T[].
-                var spanType = (NamedTypeSymbol)collectionType;
+                if (node.Elements.IsEmpty &&
+                    _compilation.GetWellKnownTypeMember(WellKnownMember.System_Collections_Immutable_ImmutableArray_T__Empty) is FieldSymbol immutableArrayOfTEmpty)
+                {
+                    // ImmutableArray<T> value = [];
+                    var immutableArrayOfTargetCollectionTypeEmpty = immutableArrayOfTEmpty.AsMember((NamedTypeSymbol)node.Type);
+                    return _factory.Field(receiver: null, immutableArrayOfTargetCollectionTypeEmpty);
+                }
 
-                Debug.Assert(collectionTypeKind is CollectionExpressionTypeKind.Span or CollectionExpressionTypeKind.ReadOnlySpan);
-                Debug.Assert(spanType.OriginalDefinition.Equals(_compilation.GetWellKnownType(
-                    collectionTypeKind == CollectionExpressionTypeKind.Span ? WellKnownType.System_Span_T : WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.AllIgnoreOptions));
-                Debug.Assert(elementType.Equals(spanType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0], TypeCompareKind.AllIgnoreOptions));
+                if (CanOptimizeSingleSpreadAsCollectionBuilderArgument(node, out _))
+                {
+                    // ReadOnlySpan<T> span = [elems];
+                    // ImmutableArray<T> array = ImmutableArray.Create(span)
+                    return VisitCollectionBuilderCollectionExpression(node);
+                }
 
-                if (elements.Length == 0)
+                const WellKnownMember wellKnownMember = WellKnownMember.System_Runtime_InteropServices_ImmutableCollectionsMarshal__AsImmutableArray_T;
+                if (_compilation.GetWellKnownTypeMember(wellKnownMember) is not MethodSymbol asImmutableArray)
```

Comment:
nit: consider inverting this (`if (...AsImmutableArray available...) { ... handle AsImmutableArray ... }`)